File: Arryas&Hashing/Encode_And_Decode_Strings/Encode_And_Decode_Strings.ts
function encode(strs: string[]): string {
  let encoded = "";
  
  for (let s of strs) {
    encoded += s.length + "#" + s;
  }
  
  return encoded;
}

function decode(str: string): string[] {
  let res = [];
  let i = 0;

  while (i < str.length) {
    let j = i;
    while (str[j] !== "#") {
      j++;
    }
    
    let length = parseInt(str.substring(i, j));
    let strDecoded = str.substring(j + 1, j + 1 + length);
    res.push(strDecoded);
    
    i = j + 1 + length;
  }

  return res;
}

// Example:
const strings = ["hello", "world", "TypeScript", "is", "awesome"];
const encodedStr = encode(strings);
console.log("Encoded:", encodedStr);

const decodedStrs = decode(encodedStr);
console.log("Decoded:", decodedStrs);
// Output:
// Encoded: 5#hello5#world10#TypeScript2#is7#awesome
// Decoded: [ 'hello', 'world', 'TypeScript', 'is', 'awesome' ]

File: Arryas&Hashing/Encode_And_Decode_Strings/Encode_And_Decode_Strings.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

class Solution {
    public:
    string encode(const vector<string>& strs) {
        string encoded = "";

        for(const auto& word: strs){
            encoded += to_string(word.size()) + "#" + word;
        }

        return encoded;
    }

    public:
    vector<string> decode(string str){
        vector<string> decoded;
        int i=0;

        while(i<str.size()){
            int j = i;
            while(str[j] != '#'){
                j++;
            }

            int len = stoi(str.substr(i,j));
            string strDecoded = str.substr(j + 1, len);
            decoded.push_back(strDecoded);

            i = j + 1 + len;
        }

        return decoded;
    }
};

int main(){
    Solution sol;

    vector<string> test = {"hello", "world", "c++", "is", "awesome"};
    string enc = sol.encode(test);
    vector<string> dec = sol.decode(enc);

    cout << "encoded:" << enc << endl;

    cout << "decoded:" << endl;
    for(const auto& word: dec){
        cout << word << endl;
    }

    return 0;
}
